module/zfs/dmu.c

```diff
@@ -1040,6 +1110,52 @@ dmu_read_by_dnode(dnode_t *dn, uint64_t offset, uint64_t size, void *buf,
 	return (dmu_read_impl(dn, offset, size, buf, flags));
 }
 
+static int
+dmu_write_abd(dnode_t *dn, uint64_t offset, uint64_t size,
+    abd_t *data, uint32_t flags, dmu_tx_t *tx)
+{
+    spa_t *spa = dn->dn_objset->os_spa;
+    dmu_buf_t **dbp;
+    int numbufs, err, i;
+    size_t off = 0;
+    zio_t *rio;
+
+    /*
+     * Direct IO must be block aligned
+     */
+     ASSERT(flags & DMU_DIRECTIO);
+     ASSERT(offset % dn->dn_datablksz == 0);
+     ASSERT(size % dn->dn_datablksze == 0);
+
+     err = dmu_buf_hold_array_by_dnode(dn, offset,
+        size, B_FALSE, FTAG, &numbufs, &dbp, 0);
+     if (err)
+         return (err);
+
+     rio = zio_root(spa, NULL, NULL, ZIO_FLAG_CANFAIL);
+
+     for (i = 0; i < numbufs; i++) {
+        dmu_buf_impl_t *db = (dmu_buf_impl_t *)dbp[i];
+        size_t dsize = dn->dn_datablksz;
+        abd_t *buf;
+
+        buf = abd_get_offset_size(data, off, dsize);
+        if (i+1 == numbufs)
```

Comment:
Yup, that is correct. In order to trigger the "wait" write I pass in the NULL pio. The idea is to make use of the calling thread instead of handing off to a task queue.